### PR TITLE
AE-3160 Improve Range, Date and fix tests

### PR DESCRIPTION
**Pull request by pgan002**
* Today, yesterday, tomorrow etc start at midnight
* This week, last week, next week, previous 10 weeks etc start on midnight on Monday
* Last 3 weeks ends now, next 3 weeks starts now
* 3 weeks ago is 1 week long ending 3 weeks ago

Fix all tests broken before and by this change. Few new tests.

Todo:
* Parameter to distinguish "did *on* Tuesday" from "will do *on* Tuesday"
* 2 weeks ago (on) Tuesday
* *Since* Tuesday
* *In* 2 weeks
* 2 weeks *from now*
* 3.5 hours ago
* Three and a half hours ago
* *until|till|by* Tuesday

### CODE REVIEW

File: tests/Range_test.py [hunk 1]
```diff
@@ -37,41 +40,45 @@ def assert_range(self, _range, expected_start, expected_end,
...
+        self.assert_range(Range('last month'),
+                          datetime(2017, 5, 1, 0, 0, 0),
+                          datetime(2017, 6, 1, 0, 0, 0))
+
+        self.assert_range(Range('previous month'),
+                          datetime(2017, 5, 1, 0, 0, 0),
+                          datetime(2017, 6, 1, 0, 0, 0))
+
+        self.assert_range(Range('next 2 months'),
+                          datetime(2017, 6, 16, 19, 37, 22),
```
Comment (dpathania1) [inline tests/Range_test.py]: why is hours, mins, secs not zero here?

File: timestring/Range.py [hunk 2]
```diff
@@ -102,46 +99,80 @@ def __init__(self, start, end=None, offset=None, start_of_week=0, tz=None, verbo
 
                         # hour
                         elif delta.startswith('h'):
-                            start = Date("today", offset=dict(hour=now.hour+1), tz=tz)
+                            start = Date("today", offset=dict(hour=now.hour), tz=tz)
```
Comment (dpathania1) [inline timestring/Range.py]: Why is now.hour+1 replaced with now.hour ?

File: tests/tests.py [hunk 3]
```diff
@@ -172,25 +170,21 @@ def test_this(self):
...
-        self.assertEqual(year.start.month, (now + timedelta(days=1)).month)
-        self.assertEqual(year.start.day, (now + timedelta(days=1)).day)
-        self.assertEqual(year.start.hour, 0)
-        self.assertEqual(year.start.minute, 0)
-        self.assertEqual(year.end.year, (now + timedelta(days=1)).year)
-        self.assertEqual(year.end.month, (now + timedelta(days=1)).month)
-        self.assertEqual(year.end.day, (now + timedelta(days=1)).day)
-        self.assertEqual(year.end.hour, 0)
-        self.assertEqual(year.end.minute, 0)
+        # 1 year (till now)
```
Comment (dpathania1) [inline tests/tests.py]: can be both from now or till now. But our music_replay use-case needs till now.

File: tests/tests.py [hunk 4]
```diff
@@ -300,27 +301,35 @@ def test_compare(self):
...
+
+        self.assertTrue(Date('last tuesday') in Range('last 14 days'))
+        self.assertTrue(Date('tuesday') in Range('next 7 days'))
+        self.assertTrue(Date('tuesday') in Range('next 7 days'))
+        self.assertTrue(Date('next tuesday') in Range('next 14 days'))
+
+    def test_ago(self):
+        now = datetime.now().replace(microsecond=0)
+        range = Range('1 year ago')
+        self.assertEqual(range.start.date.replace(microsecond=0), now.replace(year=now.year - 2))
```
Comment (dpathania1) [inline tests/tests.py]: shouldn't it be now.year-1 to now.year?

File: tests/tests.py [hunk 5]
```diff
@@ -300,27 +301,35 @@ def test_compare(self):
...
+        self.assertEqual(range.end.date.replace(microsecond=0), now.replace(year=now.year - 1))
+
+        range = Range('2 weeks ago')
+        week = timedelta(days=7)
+        self.assertEqual(range.start.date.replace(microsecond=0), now - 3 * week)
```
Comment (dpathania1) [inline tests/tests.py]: same here.

File: timestring/Range.py [hunk 6]
```diff
@@ -102,46 +99,80 @@ def __init__(self, start, end=None, offset=None, start_of_week=0, tz=None, verbo
...
 
                         else:
                             raise TimestringInvalid("Not a valid time reference")
 
                         end = start + di
 
-                    #next          x [      ]
-                    elif group['ref'] == 'next' or group['ref'] == 'upcoming':
+                    # "next 2 weeks", "the next hour"   x[     ][     ]
+                    elif group['ref'] in ['next'] and (group['num'] or group['article']):
```
Comment (dpathania1) [inline timestring/Range.py]: should it check with 'upcoming' along with 'next'?

File: tests/Range_test.py (hunk 1 above)
Comment (pgan002) [reply tests/Range_test.py]: In my understanding of English usage, next 2 months starts now. Meja agreed. Let me know if you disagree.

File: tests/tests.py (hunk 3 above)
Comment (pgan002) [reply tests/tests.py]: We have no way to know without "next" or "previous" etc. The query is just "1 year". It was in the original implementation and tests. I don't think this should be supported, but maybe they have use cases.

File: tests/tests.py (hunk 4 above)
Comment (pgan002) [reply tests/tests.py]: 1 year ago is the period that ends 1 year ago.

File: timestring/Range.py (hunk 2 above)
Comment (pgan002) [reply timestring/Range.py]: "this" and "hour" starts at `now.hour`, not `now.hour + 1`.

File: tests/tests.py (hunk 5 above)
Comment (pgan002) [reply tests/tests.py]: As above.

File: timestring/Range.py (hunk 6 above)
Comment (pgan002) [inline timestring/Range.py]: Will update.

Comment (scayetanot) [reply timestring/Range.py]: upcoming is necessary

**Comment by justonj** [comment]
@pgan002 doesn't this need to be merged?

**Review by dpathania1 — APPROVED** [review]

**Pull request closed without merging**